feat(get-started): getting started and deploy your own app (#29)
armory/docs-cdaas: content/en/get-started/deploy-your-app.md

## Learning objectives

> 

Reviewer: Add some kind of introductory statement

```suggestion
In this guide, you use a blue/green strategy to deploy your own app to your Kubernetes cluster. If you don't have a cluster, you can use [kind](https://kind.sigs.k8s.io/) to install Kubernetes locally. 

>If you prefer a web-based, interactive tutorial, see the CD-as-a-Service Console's [**Deploy Your Own App** tutorial](https://next.console.cloud.armory.io/getting-started).
```

> 1. [Create a deployment config file](#create-a-deployment-config-file)

Reviewer 2: these internal links don't auto scroll to the section.

2. [Deploy your app](#deploy-your-app)

> 3. [Monitor your deployment](#monitor-your-deployment)

Reviewer 2: There is no bigger Monitor subsection Section to link to. We do teach users to click the link CLI generated.

## {{% heading "prereq" %}}

> 

Reviewer: I think we should include a point:

```
* You are familiar with the material in the [Quickstart]({{< ref "get-started/quickstart" >}}.
```

If we don't, then we should not assume the user knows deployment file syntax, how to issue manual approvals, or where to find the deploy preview link.

> * You have the [Armory CLI installed]({{< ref "cli.md" >}}).

Reviewer 2: The page should stand on its own, Lets have collapsible section on how to install CLI, and how to install RNA.

See example here - https://github.com/armory/docs/pull/2047/files#diff-0f0bc3c26ed0b12cf6ab4fb2ddeade37fae2b10eb328d3ecce75cea1da27c17eR61

> * You have an [Armory RNA installed]({{< ref "remote-network-agent/overview.md" >}}) in your cluster.

Reviewer: [Use active voice](https://docs.armory.io/contribute/create/guides/docs-style-guide/#use-active-voice)

Parth and I talked a while ago about putting an expandable "show me how" under each of these bullet points but that's not clear in the Jira ticket.

- You have installed the Armory CLI
-- Show me how expandable section with the install command
- You have installed a Remote Network Agent installed in your cluster
-- Show me how expandable section with the CLI command

(formatting not guaranteed to have survived copy/paste)
```
* You have installed the [Armory CLI]({{< ref "cli.md" >}}).

  <details><summary>Show me how</summary>

  </details>

* You have installed a [Remote Network Agent]({{< ref "remote-network-agent/overview.md" >}}) in your cluster.

  <details><summary>Show me how</summary>
  
  Mac/Linux:

  ```bash (you might have to use HTML tags instead of backticks)
  curl -sL go.armory.io/get-cli | bash
  ```

  Windows:

  Download the latest `armory-cli` Windows executable from the [repo Releases page](https://github.com/armory-io/armory-cli/releases). Install on the machine that has access to your Kubernetes cluster.

  </details>
```

```

* You have created the Kubernetes manifests for your web app.
* You have two versions of your app to deploy.

### Directory structure

> In this guide you will create a deployment config and two simple namespace configs. The namespace manifests should be in a `manifests` directory along with the Kubernetes manifests for deploying your app.

Reviewer: ```suggestion
In this guide you create a deployment config and two simple namespace configs. The namespace manifests should be in a `manifests` directory along with the Kubernetes manifests for deploying your app.
```

Removed "will" - [Use present tense](https://docs.armory.io/contribute/create/guides/docs-style-guide/#use-present-tense)

The directory structure should look like this:

```
<your-app>

> ├── deployment.yaml

Reviewer 2: We should add a comment explaining that we are going to create this file here in this step.

└── manifests
    ├── <your-app-service>.yaml
    ├── <your-app>.yaml
    ├── namespace-staging.yaml
    └── namespace-prod.yaml
```

> ## 1. Create a deployment config file

Reviewer: We don't need to put a number in the heading. The steps are in order on the page and listed numerically in the Learning Objectives section, so that's usually enough. 
Also, removing the number will fix the link in the Learning Objectives section

```suggestion
## Create a deployment config file
```

Author: This was updated 👍

> First create two manifests for the staging and prod namespaces. These are where you'll deploy your app and also showcases the ability to deploy manifests to specific targets. Save these to the `manifests` directory.

Reviewer: ```suggestion
To illustrate deploying to different targets, you deploy to two namespaces in the same cluster. In production, you would deploy to different clusters. 

First, create two manifests, one for staging (`namespace-staging.yaml`) and the other for prod (`namespace-prod.yaml`). Save these to the `manifests` directory.
```

Use present tense

{{< cardpane >}}
{{< card code=true header="namespace-staging.yaml" lang="yaml">}}
apiVersion: v1
kind: Namespace
metadata:
  name: <your-staging-namespace>
{{< /card >}}
{{< card code=true header="namespace-prod.yaml" lang="yaml" >}}
apiVersion: v1
kind: Namespace
metadata:
  name: <your-prod-namespace>
{{< /card >}}
{{< /cardpane >}}

> Next, save the following config as `deployment.yaml` to the root of your app directory, next to the `manifests` directory. Replace the bracketed values with your own.

Reviewer 2: I think we need to guide users how to fill the bracketed values in apart from the comments in YAML. we need to make a distinction if the values are free-from logical names or if they need to match something precisely.

e.g. account needs to match the agentIdentifier, but target name is free form.

Reviewer: To follow up on Parth comment, for the namespace files, we could tell the user what to name the namespaces. For example in the staging card code:

```
apiVersion: v1
kind: Namespace
metadata:
  name: staging
```

(benefit of hindsight. I wrote this page in haste and parth approved it in haste lol)

        - pause:
            untilApproved: true
manifests:
  - path: manifests/<your-app>.yaml # replace with the name of your app manifest

>   - path: manifests/<your-app-service>.yaml # replace with the name of your app service manifest

Reviewer: good catch on the placeholders!

            weight: 100
```

> In your CDaaS config, you can create deployment strategies with as many steps as you want. In this example, the deployment config defines the following strategies:

Reviewer 2: ```suggestion
In your CD-as-a-Service config, you can create deployment strategies with as many steps as you want. In this example, the deployment config defines the following strategies:
```

* `rolling`: deploy 100% of the app (staging deployment)
* `trafficSplit`: 75% to the current version and 25% to the new version (prod deployment)

The prod deployment requires a manual approval to begin deployment and another to continue deployment after the traffic split.

> ## 2. Deploy your app

Reviewer 2: Its probably worthwhile repeating in this section that - You need to have Armory CLI installed to start a deployment with CD-as-a-Service.

> In order to see some deployment strategies at work, you'll need to deploy your app twice. The first time, CDaaS will skip any traffic splitting or manual approval steps in order to have a stable version fully deployed. The second deploy will demonstrate an advanced deployment strategy, routing only 25% of traffic to your new version, while routing the rest to the stable version.

Reviewer 2: change CDaaS to CD-as-a-Service

> In order to see some deployment strategies at work, you'll need to deploy your app twice. The first time, CDaaS will skip any traffic splitting or manual approval steps in order to have a stable version fully deployed. The second deploy will demonstrate an advanced deployment strategy, routing only 25% of traffic to your new version, while routing the rest to the stable version.

Reviewer: Use present tense https://docs.armory.io/contribute/create/guides/docs-style-guide/#use-present-tense

Reviewer: This is pretty important, so how about calling it out with an `alert` shortcode or  '>' ('>In order to...)

### Deploy the first version

1. Log in using the CLI.

   ```bash
   armory login
   ```

   The CLI returns a `Device Code` and opens your default browser.  Confirm the code in your browser to complete the login process.

2. Start the deployment from the root of your directory.

> 

Reviewer: Numercial lists should all be:
```
1.
1.
1.
```

The list is compiled to the correct numbers. Using all 1s saves you from having to renumber if you insert another step into the top or middle of the list.

   ```bash
   armory deploy start  -f deployment.yaml
   ```

>    This command starts your deployment, then returns a **Deployment ID** and a link to your deployment details. 

Reviewer: ```suggestion
   This command starts your deployment, and then returns a **Deployment ID** and a link to your deployment details. 
```

3. Monitor your deployment execution.

   Use the link provided by the CLI to view your deployment in the [CD-as-a-Service Console](https://console.cloud.armory.io/deployments). 

4. Issue manual approval.

   Once CD-as-a-Service successfully deploys your resources to `staging`, it waits for your manual approval before deploying to `prod`. When the `staging` deployment has completed, click **Approve** to allow the `prod` deployment to begin. 
   > You must issue manual approvals using the UI. You cannot issue manual approvals using the CLI.

>   Because this is the first time deploying your app, CD-as-a-Service deploys 100% to your prod environment, skipping the defined `trafficSplit` strategy. CD-as-a-Service uses the `trafficSplit` when deploying subsequent versions of your app.

Reviewer: Is this supposed to be part of #4? If yes, I think you need to add another space for the sentence to align.

![image](https://github.com/armory/docs-cdaas/assets/26799583/a75e89a3-455d-4516-9d86-b4c59b27f41f)

{{% alert title="Important" color="warning" %}}
CDaaS manages your Kubernetes deployments using ReplicaSets. During the initial deployment of your app, CDaaS deletes the underlying Kubernetes deployment object in a way that leaves behind the ReplicaSet and pods so that there is no actual downtime for your app. These are later deleted when the deployment succeeds.

> If your initial deployment fails, you should [manually delete]({{< ref "troubleshooting/tools#initial-deployment-failure-orphaned-replicaset" >}}) the orphaned ReplicaSet.

Reviewer: ```suggestion
CD-as-a-Service manages your Kubernetes deployments using ReplicaSets. During the initial deployment of your app, CD-as-a-Service  deletes the underlying Kubernetes deployment object in a way that leaves behind the ReplicaSet and pods so that there is no actual downtime for your app. These are later deleted when the deployment succeeds.

If your initial deployment fails, you should [manually delete]({{< ref "troubleshooting/tools#initial-deployment-failure-orphaned-replicaset" >}}) the orphaned ReplicaSet.
```
Can't use CDaaS as the product name. Has to be CD-as-a-Service, Armory CD-as-a-Service, or Armory Continuous Deployment-as-a-Service  https://docs.armory.io/contribute/create/guides/docs-style-guide/#armory-products-list

{{% /alert %}}

### Deploy second version

1. Update your Kubernetes manifest to deploy a new version of your app. 

> 2. Start the deployment from the root of your directory.

Reviewer: Numercial lists should all be:
```
1.
1.
1.
```

The list is compiled to the correct numbers. Using all 1s saves you from having to renumber if you insert another step into the top or middle of the list.

   ```bash
   armory deploy start  -f deployment.yaml
   ```

3. Monitor your deployment.

4. Issue manual approvals.

>    Approve the deployment to prod again and this time, observe the traffic split on the deployment details page and preview your web app deployment. Issue your manual approval to finish deployment.

Reviewer: ```suggestion
   Approve the deployment to prod and this time, observe the traffic split on the deployment details page. You can also preview your web app deployment. Issue your manual approval to finish deployment.
```
nits

## {{% heading "nextSteps" %}}

> * {{< linkWithTitle "deployment/strategies/overview.md" >}}

Reviewer 2: I think we'll benefit from adding a custom text here but link to the same pages. 

e.g. 
Learn about different deployment strategies
Use Github Action to automatically trigger deployments
Integrate your tool chain using webhooks.
